OpenAI_API/Chat/ChatEndpoint.cs

```diff
@@ -0,0 +1,199 @@
+			return CreateChatAsync(request);
+		}
+
+		/// <summary>
+		/// Ask the API to complete the request using the specified parameters. This is non-streaming, so it will wait until the API returns the full result. Any non-specified parameters will fall back to default values specified in <see cref="DefaultChatRequestArgs"/> if present.
+		/// </summary>
+		/// <param name="messages">The array of messages to send to the API</param>
+		/// <param name="model">The model to use. See the ChatGPT models available from <see cref="ModelsEndpoint.GetModelsAsync()"/></param>
+		/// <param name="temperature">What sampling temperature to use. Higher values means the model will take more risks. Try 0.9 for more creative applications, and 0 (argmax sampling) for ones with a well-defined answer. It is generally recommend to use this or <paramref name="top_p"/> but not both.</param>
+		/// <param name="top_p">An alternative to sampling with temperature, called nucleus sampling, where the model considers the results of the tokens with top_p probability mass. So 0.1 means only the tokens comprising the top 10% probability mass are considered. It is generally recommend to use this or <paramref name="temperature"/> but not both.</param>
+		/// <param name="numOutputs">How many different choices to request for each prompt.</param>
+		/// <param name="max_tokens">How many tokens to complete to. Can return fewer if a stop sequence is hit.</param>
+		/// <param name="frequencyPenalty">The scale of the penalty for how often a token is used.  Should generally be between 0 and 1, although negative numbers are allowed to encourage token reuse.</param>
+		/// <param name="presencePenalty">The scale of the penalty applied if a token is already present at all.  Should generally be between 0 and 1, although negative numbers are allowed to encourage token reuse.</param>
+		/// <param name="logitBias">Maps tokens (specified by their token ID in the tokenizer) to an associated bias value from -100 to 100. Mathematically, the bias is added to the logits generated by the model prior to sampling. The exact effect will vary per model, but values between -1 and 1 should decrease or increase likelihood of selection; values like -100 or 100 should result in a ban or exclusive selection of the relevant token.</param>
+		/// <param name="stopSequences">One or more sequences where the API will stop generating further tokens. The returned text will not contain the stop sequence.</param>
+		/// <returns>Asynchronously returns the completion result. Look in its <see cref="ChatResult.Choices"/> property for the results.</returns>
+		public Task<ChatResult> CreateChatAsync(List<ChatMessage> messages = null,
+			Model model = null,
+			double? temperature = null,
+			double? top_p = null,
+			int? numOutputs = null,
+			int? max_tokens = null,
+			double? frequencyPenalty = null,
+			double? presencePenalty = null,
+			Dictionary<string, float> logitBias = null,
```

Comment:
Should this be an `IReadOnlyDictionary` or a `IDictionary`?

Since this is in the outgoing request, `logitBias` is something the user has created themselves.
Shouldn't it be a regular dictionary since the user is the one making it?

I noted it in the PR, but I do not actually understand what [logit_bias](https://platform.openai.com/docs/api-reference/chat/create#chat/create-logit_bias) is used for, so I am not sure about it's implementation here.

Reply:
It's a good question. The TL;DR is by taking the least derived type, an `IReadOnlyDictionary<>` in this case, you make the intention of your method clearer and it's more broadly callable. The caller doesn't need to create a specific `Dictionary<>`, they can pass any type that implements the interface (which includes `Dictionary<>`, so you lose nothing)

As the Framework Design Guidelines ponit out, method arguments should be the least derived type needed by the method. This makes it more broadly callable. For example, the caller of the method might not have a `Dictionary<>` handy. They might have something that implements the interface though.

```csharp
IReadOnlyDictionary<string, float> biases = GetBiases(...);
```

Now then they want to pass the `biases` into your method, they should be able to without having to call `ToDictionary()`.

> Since this is in the outgoing request

I'd argue that `ChatRequest.LogitBias` (the outgoing request) could also be an `IReadOnlyDictionary<string, float>` (or `IReadOnlyDictionary<string, int>`, but it's not clear from the documentation). That will get serialized properly.

The only time your method needs to take in a `Dictionary<>` is if it's supposed to modify the passed in dictionary, which should be rare. And if your method. modifies the dictionary, it should be very clear to the caller that's what it does.

As a caller of an API, I'd be worried about passing in a concrete `Dictionary<>` to a method because maybe the method modifies it and I don't realize it, which could lead to bugs in my code. If I saw a method argument of `Dictionary<>`, I'd probably create a copy of my dictionary first via `ToDictionary()` to be safe so that anything your method does won't affect my dictionary.

By making the argument `IReadOnlyDictionary<>`, you let the caller know that you don't intend to modify the dictionary, but only read its values. Advertising your intentions is good API design. :smile:

Reply:
> I'd argue that `ChatRequest.LogitBias` (the outgoing request) could also be an `IReadOnlyDictionary<string, float>` (or `IReadOnlyDictionary<string, int>`, but it's not clear from the documentation). That will get serialized properly.

I should be clear, if the `ChatRequest` object is using a builder pattern, then it's fine if the property is `Dictionary<>`. But if it's meant to be created all at once, it could be a read only dictionary. This one is not that important to me. My focus is more on the method arguments.

Reply:
> As a caller of an API, I'd be worried about passing in a concrete `Dictionary<>` to a method because maybe the method modifies it and I don't realize it, which could lead to bugs in my code. If I saw a method argument of `Dictionary<>`, I'd probably create a copy of my dictionary first via `ToDictionary()` to be safe so that anything your method does won't affect my dictionary.
> 
> By making the argument `IReadOnlyDictionary<>`, you let the caller know that you don't intend to modify the dictionary, but only read its values. Advertising your intentions is good API design. 😄

That makes sense to me!

> I'd argue that ChatRequest.LogitBias (the outgoing request) could also be an IReadOnlyDictionary<string, float> (or IReadOnlyDictionary<string, int>, but it's not clear from the documentation). That will get serialized properly.

Yeah, it's not clear to me from the documentation either. I might have to look at some other implementations and see what they do. I was hoping someone else would chime in here with the correct answer eventually.